consensus/src/lib.rs

```diff
@@ -1,27 +1,1112 @@
+        match block_index {
+            Some(block_index) => {
+                let prev_block_id =
+                    block_index.get_prev_block_id().as_ref().ok_or(BlockError::NotFound)?;
+                Ok(self.db_tx.get_block_index(prev_block_id)?.ok_or(BlockError::NotFound)?)
+            }
+            None => Err(BlockError::NotFound),
+        }
+    }
+
+    // Get indexes for a new longest chain
+    fn get_new_chain(
+        &self,
+        new_tip_block_index: &BlockIndex,
+    ) -> Result<Vec<BlockIndex>, BlockError> {
+        let mut result = Vec::new();
+        let mut block_index = new_tip_block_index.clone();
+        while !self.is_block_in_main_chain(&block_index) {
+            result.insert(0, block_index.clone());
+            block_index = self.get_ancestor(block_index.get_block_id())?;
+        }
+        Ok(result)
+    }
+
+    fn reorganize(
+        &mut self,
+        best_block_id: Option<Id<Block>>,
+        top_block_index: &BlockIndex,
+    ) -> Result<(), BlockError> {
+        let new_chain = self.get_new_chain(top_block_index)?;
+
+        // Disconnect the current chain if it is not a genesis
+        if let Some(ref best_block_id) = best_block_id {
+            let common_ancestor = self.get_ancestor(new_chain[0].get_block_id())?;
+            let mut current_ancestor = self
+                .db_tx
+                .get_block_index(best_block_id)?
+                .expect("Can't get block index. Inconsistent DB");
+
+            // Disconnect blocks
+            while self.is_block_in_main_chain(&current_ancestor)
+                && current_ancestor.get_block_id() != common_ancestor.get_block_id()
+            {
+                self.disconnect_tip(&mut current_ancestor)?;
+                current_ancestor = self.get_ancestor(current_ancestor.get_block_id())?;
+            }
+        }
+
+        // Connect the new chain
+        for block_index in new_chain {
+            self.connect_tip(&block_index, &best_block_id)?;
+        }
+        Ok(())
+    }
+
+    fn store_cached_inputs(&mut self, cached_inputs: &CachedInputs) -> Result<(), BlockError> {
+        for (tx_id, tx_index) in cached_inputs.iter() {
+            self.db_tx.set_mainchain_tx_index(tx_id, tx_index)?;
+        }
+        Ok(())
+    }
+
+    fn calculate_indices(
+        &self,
+        block: &Block,
+        tx: &Transaction,
+    ) -> Result<TxMainChainIndex, BlockError> {
+        let enc_block = block.encode();
+        let enc_tx = tx.encode();
+        let offset_tx = enc_block
+            .windows(enc_tx.len())
+            .enumerate()
+            .find_map(|(i, d)| (d == enc_tx).then(|| i))
+            .ok_or(BlockError::Unknown)? as u32;
+
+        let tx_position =
+            TxMainChainPosition::new(&block.get_id().get(), offset_tx as u32, enc_tx.len() as u32);
+
+        assert_eq!(
+            &self.db_tx.get_mainchain_tx_by_position(&tx_position).ok().flatten().unwrap(),
+            tx
+        );
```

Comment:
Should this be here or should it return an error instead of panicking?

Reply:
Database is corrupted if we get error here. Most likely, we will be unable to use blockchain after that. I assume that panicking is a valid here.

Reply:
Except in the tests

Reply:
Fixed!